### PR TITLE
Document FilesPipeline.file_path and ImagesPipeline.file_path

### LINKED ISSUE #2253: Document how to customize path for written files with media pipelines
Motivation: https://github.com/scrapy/scrapy/issues/923

### PR DESCRIPTION
**Pull request by Gallaecio**
Fixes #2253

### CODE REVIEW

File: docs/topics/media-pipeline.rst [hunk 1]
```diff
@@ -392,6 +392,30 @@ See here the methods that you can override in your custom Files Pipeline:
...
+      :class:`response <scrapy.http.Response>`.
+
+      In addition to ``response``, this method receives the original
+      :class:`request <scrapy.Request>` and
+      :class:`info <scrapy.pipelines.media.MediaPipeline.SpiderInfo>`.
+
+      You can override this method to customize the download path of each file.
+
+      For example, to download all files into the ``files`` folder with their
+      original filenames::
```
Comment (kmike) [inline docs/topics/media-pipeline.rst]: I think we should be a bit more clear on that it is just an example, which works for a certain kind of URLs - `http://example.com/foo.png` would be stored properly, but e.g. all files like `http://example.com/file?name=foo.png` would be stored as "file", overwriting each other. Possible solutions:

* say that it is an example, which works on websites where file URLs follow certain rules
* make it more robust (but less readable) by adding full URL hash to the path, in addition to guessed file name.

A full implementation should probably also check Content-Disposition header, to get a file name.

File: docs/topics/media-pipeline.rst [hunk 2]
```diff
@@ -475,6 +499,30 @@ See here the methods that you can override in your custom Images Pipeline:
...
+
+        import os
+        from urllib.parse import urlparse
+
+         def file_path(self, request, response, info):
```
Comment (schjoq) [inline docs/topics/media-pipeline.rst]: Hi. I found this definition wouldn't work, without assigning default values to the keyword arguments ``response`` and ``info``:
```suggestion
         def file_path(self, request, response=None, info=None):
```
Just like in:
https://github.com/scrapy/scrapy/blob/b364bfb68b481a696141aa6b7b3312679670c38e/scrapy/pipelines/files.py#L460

Comment (Gallaecio) [reply docs/topics/media-pipeline.rst]: I’ll look into it in detail later, but if I recall right those two fields are only needed when using some deprecated methods.

Comment (Gallaecio) [reply docs/topics/media-pipeline.rst]: Verified, that is only needed to support subclasses of these pipelines that still use the deprecated `file_key` method.

File: docs/topics/media-pipeline.rst (hunk 1 above)
Comment (Gallaecio) [reply docs/topics/media-pipeline.rst]: I’ve gone with option 1, and also tried to make it obvious that the example implementation does not take subdirectories into account either.

File: docs/topics/media-pipeline.rst [hunk 3]
```diff
@@ -392,6 +392,32 @@ See here the methods that you can override in your custom Files Pipeline:
...
+      For example, if file URLs end like regular paths (e.g.
+      ``https://example.com/a/b/c/foo.png``), you can use the following
+      approach to download all files into the ``files`` folder with their
+      original filenames (e.g. ``files/foo.png``)::
+
+        import os
+        from urllib.parse import urlparse
+
+         def file_path(self, request, response, info):
+             return 'files/' + os.path.basename(urlparse(request.url).path)
```
Comment (kmike) [inline docs/topics/media-pipeline.rst]: to make it a bit more clear, could you please add some boilerplate, e.g. 

```py
# ...
class MyFilesPipeline(FilesPipeline):
    # ...
    def file_path(self, request, response, info):
```

Comment (Gallaecio) [reply docs/topics/media-pipeline.rst]: I’ve gone with the whole implementation, as it only required a few more lines.

**Comment by kmike** [comment]
Looks good, thanks @Gallaecio!

**Pull request merged**